src/PIL/Jpeg2KImagePlugin.py

```diff
@@ -33,12 +33,12 @@ def __init__(self, fp, length=-1):
         self.remaining_in_box = -1
 
     def _can_read(self, num_bytes):
+        if self.has_length and self.fp.tell() + num_bytes > self.length:
+            # Outside box: ensure we don't read past the known file length
+            return False
         if self.remaining_in_box >= 0:
```

Comment:
This could technically also be an elif for symmetry and whatnot -- not that it matters, since we return from the previous if branch anyway. 👍 on spotting and fixing the bug where we could read past the parent box length!